src/librustc_data_structures/obligation_forest/mod.rs

```diff
@@ -162,38 +206,119 @@ pub struct ObligationForest<O: ForestObligation> {
+    /// The offset that this `ObligationForest` has registered. Should be de-registered before
+    /// dropping this forest.
+    watcher_offset: Option<O::WatcherOffset>,
+    /// Reusable vector for storing unblocked nodes whose watch should be removed
+    temp_unblocked_nodes: Vec<O::Variable>,
+}
+
+/// Helper struct for use with `BinaryHeap` to process nodes in the order that they were added to
+/// the forest
+struct Unblocked {
+    index: NodeIndex,
+    order: u32,
+}
+
+impl PartialEq for Unblocked {
+    fn eq(&self, other: &Self) -> bool {
+        self.order == other.order
+    }
+}
+impl Eq for Unblocked {}
+impl PartialOrd for Unblocked {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        other.order.partial_cmp(&self.order)
+    }
+}
+impl Ord for Unblocked {
+    fn cmp(&self, other: &Self) -> Ordering {
+        other.order.cmp(&self.order)
+    }
 }
 
 #[derive(Debug)]
-struct Node<O> {
+struct Node<O: ForestObligation> {
     obligation: O,
     state: Cell<NodeState>,
 
+    /// A predicate (and its key) can change during processing. If it does we need to register the
+    /// old predicate so that we can remove or mark it as done if this node errors or is done.
+    alternative_predicates: Vec<O::CacheKey>,
+
     /// Obligations that depend on this obligation for their completion. They
     /// must all be in a non-pending state.
-    dependents: Vec<usize>,
+    dependents: Vec<NodeIndex>,
+    /// Obligations that this obligation depends on for their completion.
+    reverse_dependents: Vec<NodeIndex>,
 
     /// If true, `dependents[0]` points to a "parent" node, which requires
     /// special treatment upon error but is otherwise treated the same.
     /// (It would be more idiomatic to store the parent node in a separate
-    /// `Option<usize>` field, but that slows down the common case of
+    /// `Option<NodeIndex>` field, but that slows down the common case of
     /// iterating over the parent and other descendants together.)
     has_parent: bool,
 
     /// Identifier of the obligation tree to which this node belongs.
     obligation_tree_id: ObligationTreeId,
+    node_number: u32,
 }
 
-impl<O> Node<O> {
-    fn new(parent: Option<usize>, obligation: O, obligation_tree_id: ObligationTreeId) -> Node<O> {
+impl<O> Node<O>
+where
+    O: ForestObligation,
+{
+    fn new(
+        parent: Option<NodeIndex>,
+        obligation: O,
+        obligation_tree_id: ObligationTreeId,
+        node_number: u32,
+    ) -> Node<O> {
         Node {
             obligation,
             state: Cell::new(NodeState::Pending),
+            alternative_predicates: vec![],
             dependents: if let Some(parent_index) = parent { vec![parent_index] } else { vec![] },
+            reverse_dependents: vec![],
             has_parent: parent.is_some(),
             obligation_tree_id,
+            node_number,
         }
     }
+
+    /// Initializes a node, reusing the existing allocations
```

Comment:
Can we give a bit more context? I would say something like

Initializes a node, reusing the existing allocations. Used when removing
a node from the `dead_nodes` list.